common-ui-compose/src/main/java/app/tivi/common/compose/insets.kt

```diff
@@ -14,64 +14,410 @@
+}
+
+/**
+ * Updates our mutable state backed [Insets] from an Android system insets.
+ */
+private fun Insets.updateFrom(insets: androidx.core.graphics.Insets) {
+    left = insets.left
+    top = insets.top
+    right = insets.right
+    bottom = insets.bottom
+}
 
-    Providers(InsetsAmbient provides currentInsets.value) {
-        children()
+/**
+ * Declare the height of the content to match the height of the status bar exactly.
+ *
+ * This is very handy when used with `Spacer` to push content below the status bar:
+ * ```
+ * Column {
+ *     Spacer(Modifier.statusBarHeight())
+ *
+ *     // Content to be drawn below status bar (y-axis)
+ * }
+ * ```
+ *
+ * It's also useful when used to draw a scrim which matches the status bar:
+ * ```
+ * Spacer(
+ *     Modifier.statusBarHeight()
+ *         .fillMaxWidth()
+ *         .drawBackground(MaterialTheme.colors.background.copy(alpha = 0.3f)
+ * )
+ * ```
+ *
+ * Internally this uses [Modifier.height] so has the same characteristics with regards to incoming
+ * layout constraints.
+ */
+fun Modifier.statusBarHeight() = composed {
+    // TODO: Move to Android 11 WindowInsets APIs when they land in AndroidX.
+    // It currently assumes that status bar == top which is probably fine, but doesn't work
+    // in multi-window, etc.
+    InsetsSizeModifier(insets = DisplayInsetsAmbient.current.systemBars, heightSide = VerticalSide.Top)
+}
+
+/**
+ * Declare the preferred height of the content to match the height of the navigation bar when present at the bottom of the screen.
+ *
+ * This is very handy when used with `Spacer` to push content below the navigation bar:
+ * ```
+ * Column {
+ *     // Content to be drawn above status bar (y-axis)
+ *     Spacer(Modifier.navigationBarHeight())
+ * }
+ * ```
+ *
+ * It's also useful when used to draw a scrim which matches the navigation bar:
+ * ```
+ * Spacer(
+ *     Modifier.navigationBarHeight()
+ *         .fillMaxWidth()
+ *         .drawBackground(MaterialTheme.colors.background.copy(alpha = 0.3f)
+ * )
+ * ```
+ *
+ * Internally this uses [Modifier.height] so has the same characteristics with regards to incoming
+ * layout constraints.
+ */
+fun Modifier.navigationBarHeight() = composed {
+    // TODO: Move to Android 11 WindowInsets APIs when they land in AndroidX.
+    // It currently assumes that nav bar == bottom, which is wrong in landscape.
+    // It also doesn't handle the IME correctly.
+    InsetsSizeModifier(insets = DisplayInsetsAmbient.current.systemBars, heightSide = VerticalSide.Bottom)
+}
+
+enum class HorizontalSide { Left, Right }
+enum class VerticalSide { Top, Bottom }
+
+/**
+ * Declare the preferred width of the content to match the width of the navigation bar,
+ * on the given [side].
+ *
+ * This is very handy when used with `Spacer` to push content inside from any vertical
+ * navigation bars (typically when the device is in landscape):
+ * ```
+ * Row {
+ *     Spacer(Modifier.navigationBarWidth(HorizontalSide.Left))
```

Comment:
How would this handle RTL?

Reply:
Doesn't need to, insets are always absolute sides